src/torrent.cpp

```diff
@@ -8684,26 +8690,35 @@ namespace {
 		m_ses.deferred_submit_jobs();
 	}
 
-	void torrent::move_storage(std::string const& save_path, move_flags_t const flags)
+	void torrent::move_storage(std::string const& save_path, move_flags_t const flags, std::shared_ptr<std::promise<const storage_moved_alert*>> promise)
 	{
 		TORRENT_ASSERT(is_single_thread());
 		INVARIANT_CHECK;
 
 		if (m_abort)
 		{
-			if (alerts().should_post<storage_moved_failed_alert>())
-				alerts().emplace_alert<storage_moved_failed_alert>(get_handle()
+			if (alerts().should_post<storage_moved_failed_alert>()) {
+				auto* a = alerts().emplace_alert<storage_moved_failed_alert>(get_handle()
 					, boost::asio::error::operation_aborted
 					, "", operation_t::unknown);
+
+				if (promise) promise->set_exception(std::make_exception_ptr(alert_exception(a)));
```

Comment:
`a` is a non-owning pointer to the alert, and the life-time of the alert is tied to when the user calls `pop_alerts()`. This seems really risky as the future may be checked *after* the alert object has been deleted.